Resources/ProtobufDumper/ProtobufDumper/Program.cs

```diff
@@ -11,66 +11,108 @@ static void Main( string[] args )
+            for ( var i = 0; i < unnamedArgs.Count; i++ )
             {
-                if ( i == 0 || i < args.Length - 1 )
+                if ( i == 0 || i < unnamedArgs.Count - 1 )
                 {
-                    targets.Add( args[ i ] );
+                    targets.Add( unnamedArgs[ i ] );
                 }
                 else
                 {
-                    output = args[ i ];
+                    output = unnamedArgs[ i ];
                 }
             }
 
             var outputDir = output ?? Path.GetFileNameWithoutExtension( targets[ 0 ] );
 
-            ProtobufCollector collector = new ProtobufCollector();
+            var collector = new ProtobufCollector();
 
             foreach ( var target in targets )
             {
+                Console.WriteLine( "Loading binary '{0}'...", target );
+
                 ExecutableScanner.ScanFile( target, ( name, buffer ) =>
                 {
-                    if ( Environment.GetCommandLineArgs().Contains( "-dump", StringComparer.OrdinalIgnoreCase ) )
+                    Console.Write( "{0}... ", name );
+
+                    var complete = collector.TryParseCandidate( name, buffer, out var result, out var error );
+
+                    switch ( result )
+                    {
+                        case ProtobufCollector.CandidateResult.OK:
+                            Console.ForegroundColor = ConsoleColor.Green;
+                            Console.WriteLine( "OK!" );
+                            Console.ResetColor();
+                            break;
+
+                        case ProtobufCollector.CandidateResult.Rescan:
+                            Console.ForegroundColor = ConsoleColor.Cyan;
+                            Console.WriteLine( "needs rescan: {0}", error.Message );
+                            Console.ResetColor();
+                            break;
+
+                        default:
+                        case ProtobufCollector.CandidateResult.Invalid:
+                            Console.ForegroundColor = ConsoleColor.Red;
+                            Console.WriteLine( "is invalid: {0}", error.Message );
+                            Console.ResetColor();
+                            break;
+                    }
+
+                    if ( complete && hasDumpCandidates )
                     {
-                        Directory.CreateDirectory( outputDir );
                         var fileName = Path.Combine( outputDir, $"{name}.dump" );
                         Directory.CreateDirectory( Path.GetDirectoryName( fileName ) );
 
                         Console.WriteLine( "  ! Dumping to '{0}'!", fileName );
 
                         try
                         {
-                            File.WriteAllBytes( fileName, buffer );
+                            using ( var file = File.OpenWrite( fileName ) )
+                            {
+                                file.Write( buffer );

```

Comment:
If we’re expecting to overwrite an existing file, we should call FileStream.SetLength.